File: Python/problem014.py
#! /usr/bin/env python3
"""
Problem 14 - Project Euler
http://projecteuler.net/index.php?section=problems&id=014
"""
chains = {}


def getCollatzChainLength(n):
    chain = []
    extra = 0
    while n != 1:
        if n in chains:
            extra = chains[n]
            break
        else:
            chain.append(n)
            if n % 2 == 0:  # even
                n = n // 2
            else:  # odd
                n = n * 3 + 1
    else:
        chain.append(n)

    for i, item in enumerate(chain):
        if item not in chains:
            chains[item] = len(chain[i:]) + extra

    return len(chain) + extra


if __name__ == "__main__":
    print(
        max(
            [(x, getCollatzChainLength(x)) for x in range(1, 1_000_000)],
            key=(lambda x: x[1]),
        )
    )

File: Clojure/problem014.clj
#! /usr/bin/env clojure
; Problem 14 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=014

(defn chain' [n i]
  (if (= n 1)
    i
    (if (even? n)
      (recur (/ n 2) (inc i))
      (recur (inc (* 3 n)) (inc i)))))

(defn chain [n]
  (list n (chain' n 1)))

(->> (range 2 1000000)
     (map chain)
     (apply max-key #(second %))
     (println))
